src/Zio/FileSystems/ZipArchiveFileSystem.cs

```diff
@@ -0,0 +1,1034 @@
+                }
+            }
+            finally
+            {
+                _entriesLock.ExitWriteLock();
+            }
+        }
+
+        private ZipArchiveEntry CreateEntry(string path)
+        {
+            path = RemoveLeadingSlash(path);
+#if NET45
+            path = path.Replace('/', DirectorySeparator);
+#else
+            path = path.Replace('\\', DirectorySeparator);
+#endif
+            _entriesLock.EnterWriteLock();
+            try
+            {
+                var entry = _archive.CreateEntry(path);
+
+                if (!_isCaseSensitive)
+                {
+                    _entries.Add(entry.FullName.ToLowerInvariant(), entry);
+                }
+
+                return entry;
+            }
+            finally
+            {
+                _entriesLock.ExitWriteLock();
+            }
+        }
+
+        private string GetName(ZipArchiveEntry entry)
+        {
+            var name = entry.FullName.TrimEnd('/', '\\');
+            var index = name.LastIndexOfAny(new[] { '/', '\\' });
+            return name.Substring(index + 1);
+        }
+
+        private static string GetParent(string path)
+        {
+            path = path.TrimEnd('/', '\\');
+            var lastIndex = path.LastIndexOfAny(new[] { '/', '\\' });
+            return lastIndex == -1 ? "" : path.Substring(0, lastIndex);
+        }
+
+        private static string ConvertPathToDirectory(UPath path)
+        {
+            return ConvertPathToDirectory(path.FullName);
+        }
+
+        private static string ConvertPathToDirectory(string path)
+        {
+            if (string.IsNullOrEmpty(path))
+            {
+                return path;
+            }
+
+            return path[path.Length - 1] is DirectorySeparator ? path : path + DirectorySeparator;
+        }
+
+        private static string RemoveLeadingSlash(UPath path)
+        {
+            return path.FullName[0] is '\\' or '/' ? path.FullName.Substring(1) : path.FullName;
+        }
+
+        private static string RemoveLeadingSlash(string path)
+        {
+            return path[0] is '\\' or '/' ? path.Substring(1) : path;
+        }
+
+        private FileSystemEventDispatcher<FileSystemWatcher>? TryGetDispatcher()
+        {
+            lock (_dispatcherLock)
+            {
+                return _dispatcher;
+            }
+        }
+
+
+        private class ZipEntryStream : Stream
+        {
+            private readonly ZipArchiveEntry _entry;
+            private readonly ZipArchiveFileSystem _fileSystem;
+            private readonly Stream _streamImplementation;
+            private bool _isDisposed;
+
+            public ZipEntryStream(FileShare share, ZipArchiveFileSystem system, ZipArchiveEntry entry)
+            {
+                _entry = entry;
+                _fileSystem = system;
+                var fileShare = _fileSystem._openStreams.TryGetValue(entry, out var fileData) ? fileData.First : FileShare.ReadWrite;
```

Comment:
It seems that if `ZipArchiveFileSystem` is used from multiple threads, `_openStreams` is not protected, even just when reading, that could be a problem no?